php-checks/src/main/java/org/sonar/php/checks/phpunit/NoAssertionInTestCheck.java

```diff
@@ -0,0 +1,163 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.php.checks.phpunit;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.sonar.check.Rule;
+import org.sonar.php.checks.utils.CheckUtils;
+import org.sonar.php.checks.utils.PhpUnitCheck;
+import org.sonar.php.tree.TreeUtils;
+import org.sonar.plugins.php.api.symbols.Symbol;
+import org.sonar.plugins.php.api.symbols.SymbolTable;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.declaration.MethodDeclarationTree;
+import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonar.plugins.php.api.tree.expression.MemberAccessTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+@Rule(key = "S2699")
+public class NoAssertionInTestCheck extends PhpUnitCheck {
+  private static final String MESSAGE = "Add at least one assertion to this test case.";
+
+  private static final Pattern ASSERTION_METHODS_PATTERN = Pattern.compile("(assert|verify|fail|pass|should|check|expect|validate|test|.*Test).*");
+  private static final List<String> TEST_CONTROL_FUNCTIONS = ImmutableList.of(
+    "addtoassertioncount");
+
+  private final Map<MethodDeclarationTree, Boolean> assertionInMethod = new HashMap<>();
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    if (!isTestCaseMethod(tree)) {
+      return;
+    }
+
+    if (CheckUtils.hasAnnotation(tree, "expectedException") ||
+      CheckUtils.hasAnnotation(tree, "doesNotPerformAssertions") ||
+      CheckUtils.hasAnnotation(tree, "expectedDeprecation")) {
+      return;
+    }
+
+    AssertionsFindVisitor assertionsFindVisitor = new AssertionsFindVisitor(context().symbolTable());
+    tree.accept(assertionsFindVisitor);
+
+    if (!assertionsFindVisitor.didFindAssertion) {
+      context().newIssue(this, tree.name(), MESSAGE);
+    }
+  }
+
+  private class AssertionsFindVisitor extends PHPVisitorCheck {
+    private boolean didFindAssertion = false;
+    private final SymbolTable symbolTable;
+
+    private AssertionsFindVisitor(SymbolTable symbolTable) {
+      this.symbolTable = symbolTable;
+    }
+
+    @Override
+    public void visitFunctionCall(FunctionCallTree tree) {
+      String functionName = getFunctionName(tree);
+
+      if (isAssertion(tree) ||
+        functionNameCountsAsAssertion(functionName) ||
+        isDynamicFunctionCall(tree) ||
+        isLocalMethodWithAssertion(tree)) {
+        didFindAssertion = true;
+      }
+
+      super.visitFunctionCall(tree);
```

Comment:
Is this call necessary or superfluous?

Reply:
Necessary, when we have a chain (example `$logger->expects($this->exactly(2))->method('debug');`) the sub-parts need to be visited too.